File: practice/data_structures/statcks/largest_rectangle/largest_rectangle.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <stack>

using namespace std;

long long max_area(const vector<long long> & i_heights)
{
    stack<long long> st;
    long long res = 0;
    
    int pos = 0;
    while (pos < i_heights.size())
    {
        if (st.empty() || (i_heights[st.top()] <= i_heights[pos]))
        {
            st.push(pos);
            pos++;
        }
        else
        {
            long long top = st.top();
            st.pop();
            long long w;
            if (st.empty())
            {
                w = pos;
            }
            else
            {
                w = pos - st.top() - 1;
            }
            res = max(res, i_heights[top] * w);
        }
    }
    
    while (!st.empty())
    {
        long long top = st.top();
        st.pop();
        long long w;
        if (st.empty())
        {
            w = pos;
        }
        else
        {
            w = pos - st.top() - 1;
        }
        res = max(res, i_heights[top] * w);
    }
    
    return res;
}

int main()
{
    int n;
    cin >> n;
    
    vector<long long> heights(n);
    for (int pos = 0; pos < n; ++pos)
    {
        cin >> heights[pos];
    }
    
    cout << max_area(heights) << endl;
    return 0;
}

File: practice/data_structures/stacks/largest_rectangle/largest_rectangle.py
#!/usr/bin/env python

def max_area(i_heights):
    stack = []
    res = 0
    pos = 0
    
    while pos < len(i_heights):
        if not stack or i_heights[stack[-1]] <= i_heights[pos]:
            stack.append(pos)
            pos += 1
        else:
            top = stack.pop()
            w = pos if not stack else pos - stack[-1] - 1
            area_with_top = i_heights[top] * w
            res = max(res, area_with_top)
    while stack:
        top = stack.pop()
        w = pos if not stack else pos - stack[-1] - 1
        area_with_top = i_heights[top] * w
        res = max(res, area_with_top)
    return res

n = int(raw_input().strip())
h = map(int, raw_input().strip().split())
print max_area(h)

